### PR TITLE
Systemd Watchdog support for AVBStreamHandler

### PR DESCRIPTION
**Pull request by tvyas**

### CODE REVIEW

File: CMakeLists.avb_watchdog.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+
+#------------------------------------------------------------------
+# Build the AVB Streamhandler Library
```
Comment (edersondisouza) [inline CMakeLists.avb_watchdog.txt]: Minor nitpick: this actually builds the avb sh watchdog lib.

File: private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
+/*
+ @COPYRIGHT_TAG@
+ */
```
Comment (edersondisouza) [inline private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp]: Change this for the real license header, please.

File: private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+/**
+ * @file
```
Comment (edersondisouza) [inline private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp]: If you are not going to describe this file in more detail, just remove this comment.

File: private/inc/avb_watchdog/IasWatchdogResult.hpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/*
+@COPYRIGHT_TAG@
+*/
```
Comment (edersondisouza) [inline private/inc/avb_watchdog/IasWatchdogResult.hpp]: License =D

File: private/inc/avb_watchdog/IasWatchdogThread.hpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
+/*
+ @COPYRIGHT_TAG@
+ */
```
Comment (edersondisouza) [inline private/inc/avb_watchdog/IasWatchdogThread.hpp]: License =D

File: private/src/avb_streamhandler/IasAlsaWorkerThread.cpp [hunk 6]
```diff
@@ -447,6 +483,19 @@ IasResult IasAlsaWorkerThread::run()
     else
     {
       double timeOffset = 0.0;
+
+      if (mWatchdog != NULL) {
+	if (mWatchdog->isRegistered()) {
```
Comment (edersondisouza) [inline private/src/avb_streamhandler/IasAlsaWorkerThread.cpp]: There's an indentation issue here.

File: private/src/avb_watchdog/IasWatchdogInterface.cpp [hunk 7]
```diff
@@ -0,0 +1,307 @@
+/*
+ @COPYRIGHT_TAG@
+ */
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasWatchdogInterface.cpp]: License =D

File: private/src/avb_watchdog/IasWatchdogInterface.cpp [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+        mDltContext,
+        DLT_LOG_ERROR,
+        "Registering watchdog", watchdogName, " with timeout", timeout, " failed. Watchdog already registered", *this);
+    result = IasWatchdogResult::cAlreadyRegistered;
+  }
+  else
+  {
+    std::lock_guard<std::mutex> mutexGuard(mDueResetTimeMutex);
+
+    if ( IasMediaTransportAvb::IAS_SUCCEEDED(IasMediaTransportAvb::IasResult::cOk) )
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasWatchdogInterface.cpp]: This is also an `if (true)` case. So this if can go away along with the else clause. Only the code inside this if should remain.

File: private/src/avb_watchdog/IasWatchdogInterface.cpp [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        DLT_LOG_ERROR,
+        "Unregistering watchdog failed. Watchdog not registered"
+        "current PID=", static_cast<int32_t>(getpid()), "currentThreadId=", static_cast<uint64_t>(pthread_self()));
+    result = IasWatchdogResult::cWatchdogUnregistered;
+  }
+  else
+  {
+    std::lock_guard<std::mutex> mutexGuard(mDueResetTimeMutex);
+
+    if ( IasMediaTransportAvb::IAS_SUCCEEDED(IasMediaTransportAvb::IasResult::cOk) )
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasWatchdogInterface.cpp]: Ditto regarding the `if (true)`. The mutex locking will simply throw an exception if it fails, no other case available.

File: private/src/avb_watchdog/IasWatchdogInterface.cpp [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+uint64_t IasWatchdogInterface::getDueResetTime() const
+{
+  // Select a time that differs from zero to ensure valid timestamp if mutex lock fails.
+  uint64_t dueResetTime  = 0u;
+  dueResetTime = dueResetTime + 1000000; //1ms
+  std::lock_guard<std::mutex> mutexGuard(mDueResetTimeMutex);
+
+  if ( IasMediaTransportAvb::IAS_SUCCEEDED(IasMediaTransportAvb::IasResult::cOk) )
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasWatchdogInterface.cpp]: Another case of `if (true)`.

File: private/src/avb_watchdog/IasWatchdogInterface.cpp [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+    result = IasWatchdogResult::cIncorrectThread;
+  }
+  else
+  {
+    uint64_t const currentTime = IasWatchdogInterface::getCurrentRawTime();
+    if ( (currentTime < mDueResetTime) || (mDueResetTime==0u) )
+    {
+      std::lock_guard<std::mutex> mutexGuard(mDueResetTimeMutex);
+
+      if ( IasMediaTransportAvb::IAS_SUCCEEDED(IasMediaTransportAvb::IasResult::cOk) )
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasWatchdogInterface.cpp]: These `if (true)` keep appearing...
Note that this mutex apparently can be entirely removed by making `mDueResetTime` a `std::atomic`. But no need to do this now, just keep track of it.

File: private/src/avb_watchdog/IasWatchdogResult.cpp [hunk 12]
```diff
@@ -0,0 +1,86 @@
+/*
+@COPYRIGHT_TAG@
+*/
```
Comment (edersondisouza) [inline private/src/avb_watchdog/IasWatchdogResult.cpp]: License header

File: private/src/avb_streamhandler/IasAvbTransmitSequencer.cpp [hunk 13]
```diff
@@ -217,7 +217,7 @@ IasAvbProcessingResult IasAvbTransmitSequencer::init(uint32_t queueIndex, IasAvb
   if (eIasAvbProcOK == result)
   {
     uint64_t val = 0u;
-    (void) IasAvbStreamHandlerEnvironment::getConfigValue(IasRegKeys::cXmitUseWatchdog, val);
+    (void) IasAvbStreamHandlerEnvironment::getConfigValue(IasRegKeys::cUseWatchdog, val);
```
Comment (edersondisouza) [inline private/src/avb_streamhandler/IasAvbTransmitSequencer.cpp:220]: Maybe squash this change with previous one? So we have more commits that are build-able? This can help `git bisect` usage in the future.

File: private/src/avb_streamhandler/IasAlsaWorkerThread.cpp [hunk 14]
```diff
@@ -25,7 +25,7 @@
 namespace IasMediaTransportAvb {
 
 static const std::string cClassName = "IasAlsaWorkerThread::";
-#define LOG_PREFIX cClassName + __func__ + "(" + std::to_string(__LINE__) + "):"
```
Comment (edersondisouza) [inline private/src/avb_streamhandler/IasAlsaWorkerThread.cpp]: This change appears unrelated...

File: private/src/avb_streamhandler/IasAlsaWorkerThread.cpp [hunk 15]
```diff
@@ -97,6 +98,29 @@ IasAvbProcessingResult IasAlsaWorkerThread::init(IasAlsaStreamInterface *alsaStr
           DLT_LOG_CXX(*mLog, DLT_LOG_INFO, LOG_PREFIX, "Clock domain (",
                       uint16_t(mClockDomain->getType()), ") and Stream (",
                       alsaStream->getStreamId(), ") added");
+
+		uint64_t val = 0u;
+		(void) IasAvbStreamHandlerEnvironment::getConfigValue(IasRegKeys::cUseWatchdog, val);
+		if (val && IasAvbStreamHandlerEnvironment::isWatchdogEnabled()) {
```
Comment (edersondisouza) [inline private/src/avb_streamhandler/IasAlsaWorkerThread.cpp]: I know the project has a rather peculiar indentation, but I'd try to be consistent with it. So, something like
```
 if (true)
 {
   (...)
 }
 else
 {
    // Two spaces
 }
```

File: private/src/avb_streamhandler/IasAvbReceiveEngine.cpp [hunk 16]
```diff
@@ -159,6 +160,33 @@ IasAvbProcessingResult IasAvbReceiveEngine::init()
       result = openReceiveSocket();
     }
 
+    if (result == eIasAvbProcOK) {
+        result = eIasAvbProcInitializationFailed;
+
+	uint64_t val = 0u;
+	(void) IasAvbStreamHandlerEnvironment::getConfigValue(IasRegKeys::cUseWatchdog, val);
+	if (val && IasAvbStreamHandlerEnvironment::isWatchdogEnabled()) {
```
Comment (edersondisouza) [inline private/src/avb_streamhandler/IasAvbReceiveEngine.cpp]: If we're not configured to use watchdog, this block will return `eIasAvbProcInitializationFailed`, and that looks a problem.

File: private/inc/avb_watchdog/IasWatchdogInterface.hpp [hunk 17]
```diff
@@ -0,0 +1,187 @@
+/*
+ @COPYRIGHT_TAG@
+ */
```
Comment (edersondisouza) [inline private/inc/avb_watchdog/IasWatchdogInterface.hpp]: License header.

File: private/src/avb_streamhandler/IasAvbReceiveEngine.cpp [hunk 18]
```diff
@@ -964,6 +995,16 @@ IasResult IasAvbReceiveEngine::run()
         (void) dispatchPacket(it->second, NULL, 0u, now);
       }
       (void) unlock();
+
+	if (mWatchdog) {
```
Comment (edersondisouza) [inline private/src/avb_streamhandler/IasAvbReceiveEngine.cpp]: Hint: maybe comment why each reset/register/unregister of the watchdog takes place on different places of method `run()`, just like in the transmitter? This way future people can know why the watchdog control is necessary in each place, so code can be properly changed.

File: private/src/avb_streamhandler/IasAlsaWorkerThread.cpp (hunk 14 above)
Comment (tvyas) [reply private/src/avb_streamhandler/IasAlsaWorkerThread.cpp]: Yea that's a leftover from my experiments. I'll remove it.

File: private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp (hunk 2 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp]: My bad

File: private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp (hunk 3 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasSystemdWatchdogManager.hpp]: Yea, sorry about that.

File: private/inc/avb_watchdog/IasWatchdogResult.hpp (hunk 4 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasWatchdogResult.hpp]: Yep

File: private/inc/avb_watchdog/IasWatchdogThread.hpp (hunk 5 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasWatchdogThread.hpp]: Yep

File: private/inc/avb_watchdog/IasWatchdogInterface.hpp (hunk 17 above)
Comment (tvyas) [reply private/inc/avb_watchdog/IasWatchdogInterface.hpp]: Right

File: private/src/avb_streamhandler/IasAvbReceiveEngine.cpp (hunk 18 above)
Comment (tvyas) [reply private/src/avb_streamhandler/IasAvbReceiveEngine.cpp]: Yes, I added these based on the way they did it for the transmit thread. Will explain them.

File: private/src/avb_streamhandler/IasAvbTransmitSequencer.cpp (hunk 13 above)
Comment (tvyas) [reply private/src/avb_streamhandler/IasAvbTransmitSequencer.cpp:220]: Right

File: private/src/avb_streamhandler/IasAlsaWorkerThread.cpp (hunk 15 above)
Comment (tvyas) [reply private/src/avb_streamhandler/IasAlsaWorkerThread.cpp]: Agreed

**Comment by tvyas** [comment]
Sending a new pull request with review comments addressed
